### DISCUSSION TITLE
Keybinds to Scripts

### DISCUSSION THREAD

**Discussion by austinliuigi** [spakin/SimpInkScr#108] [Q&A]
Hi,

I’m not sure if this is a restriction of Inkscape itself, but is it possible to create custom keybindings to run SimpInkScr scripts?

Thank you for the amazing extension!

**Comment by spakin** [comment]
I could be mistaken, but I don't believe this is possible in current versions of Inkscape.  As far as I can tell, _Edit_ → _Preferences_ → _Interface_ → _Keyboard_ lets you assign keys to "Previous Extension" and "Previous Extension Settings" but not to a particular Extension (such as Simple Inkscape Scripting) or using specified parameters.

I use the "Previous Extension" key (Alt-Q on my Linux installation) quite extensively to repeatedly run Simple Inkscape Scripting scripts, but that doesn't help if you need to switch frequently among multiple scripts.

**Comment by austinliuigi** [comment]
Would it be possible by direclty editing the user keyboard xml file (`~/.config/inkscape/keys/default.xml` on linux)? There are some possible actions are can't be bounded to using the keyboard shortcut preferences window. For instance, to align selected objects horizontally along their center *relative to the first selected object*, you can bind a key to `app.object-align('hcenter first')`. This exact binding isn't available in the keyboard shortcut preferences.

I was wondering if Simple Inkscape Scripting exposes actions like these that can be bound to. Forgive me if that isn't how extensions work; I am inexperienced with inkscape development and haven't looked into how extensions work exactly.

**Reply by spakin** [reply]
It sure sounds like you know more about this than I!

The corresponding binding for Simple Inkscape Scripting should be `org.pakin.filter.simple-inkscape-scripting`, but I haven't managed to get that to work—even with no arguments.  If my `defaults.xml` file looks like
```XML
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<keys
   name="User Shortcuts">
  <bind
     gaction="org.pakin.filter.simple-inkscape-scripting.noprefs"
     keys="dollar" />
</keys>
```
Inkscape complains at startup,
```
Shortcuts::add_shortcut: No Action for org.pakin.filter.simple-inkscape-scripting.noprefs
```

I suspect that Inkscape may _first_ be loading keyboard bindings and _then_ loading extensions, but that's just a guess.

**Reply by austinliuigi** [reply]
Hmm, interesting. Thanks for testing it out and giving me a lead what I can try out. I’ll ask on the inkscape forums to see if there’s anything we can do :)

**Reply by austinliuigi** [reply]
Just to keep you posted, I filed a bug report on the Inkscape gitlab repo [here](https://gitlab.com/inkscape/inbox/-/issues/9396).

**Reply by spakin** [reply]
Thanks for filing that bug report.  I hope the Inkscape developers are able to fix this shortcoming of key bindings.

**Comment by austinliuigi** [comment]
Hey @spakin! I just saw an post on the bug report which suggested to try prepending `app.` to the gaction string. This works! At least in the sense that there's no longer an error :)

But pressing the keybinding doesn't seem to do anything. Is it expected to bring up the Simple Inkscape Scripting window? I don't exactly know how inkscape actions work so I'm not sure what the action alone does without any specified parameters, or what parameters I can use if any.

**Reply by spakin** [reply]
Drop the `.noprefs` and see what happens.  `.noprefs` means "Don't prompt for extension preferences, but use the previous ones."  It's like _Extensions_ → _Previous Extension_ as opposed to _Extensions_ → _Previous Extension Settings…_.

If dropping `.noprefs` brings up the Simple Inkscape Scripting window, the next step is to try passing parameters.  I don't know what form these should take.  On the command line, one would specify `--program="`〈_code_〉`"` to provide a short piece of Simple Inkscape Scripting code or `--py-source=`〈_file_〉 to point to a script file.

Unfortunately, it's quite likely that there's no means of passing arguments to an extension (as opposed to passing arguments to a built-in action) .  At the time of this writing, Inkscape issue [#4106](https://gitlab.com/inkscape/inkscape/-/issues/4106), "Passing arguments to extensions", which I opened more than half a year ago, is still open.
